docs/source/design/result_records.rst

```diff
@@ -0,0 +1,209 @@
+record with additional information that improves its interpretability, or
+triggers particular optional functionality in generic result processing.
+
+
+``type``
+--------
+
+This field indicates the type of entity a result is associated with. This may
+or may not be the type of the local entity identified by the ``path`` value.
+The following values are common, and should be used in matching cases, but
+arbitrary other values are supported too:
+
+- ``dataset``: a DataLad dataset
+- ``file``: a regular file
+- ``directory``: a directory
+- ``symlink``: a symbolic link
+- ``key``: a git-annex key
+- ``sibling``: a Dataset sibling or Git remote
+
+
+``message``
+-----------
+
+A message providing additional human-readable information on the nature or
+provenance of a result. Any non-``ok`` results *should* have a message providing
+information on the rational of their status characterization.
+
+A message can be a string or a tuple. In case of a tuple, the second item can
+contain values for ``%``-expansion of the message string. Expansion is performed
+only immediately prior to actually outputting the message, hence string formatting
+runtime costs can be avoided this way, if a message is not actually shown.
+
+
+``logger``
+----------
+
+If a result record has a ``message`` field, then a given `Logger` instance
+(typically from ``logging.getLogger()``) will be used to automatically log
+this message. The log channel/level is determined based on
+``datalad.log.result-level`` configuration setting. By default, this is
+the ``debug`` level. When set to ``match-status`` the log level is determined
+based on the ``status`` field of a result record:
+
+- ``debug`` for ``'ok'``, and ``'notneeded'`` results
+- ``warning`` for ``'impossible`` results
+- ``error`` for ``'impossible`` results
+
+This feature should be used with care. Unconditional logging can lead to
+confusing double-reporting when results rendered and also visibly logged.
+
+
+``refds``
+---------
+
+This field can identify a path (using the same semantics and requirements as
+the ``path`` field) to a reference dataset that represents the larger context
+of an operation. For example, when recursively processing multiple files across
+a number of subdatasets, a ``refds`` value may point to the common superdataset.
+This value may influence, for example, how paths are rendered in user-output.
+
+
+``parentds``
+------------
+
+This field can identify a path (using the same semantics and requirements as
+the ``path`` field) to a dataset containing an entity.
+
+
+``state``
+---------
+
+A string label categorizing the state of an entity. Common values are:
+
+- ``clean``
+- ``untracked``
+- ``modified``
+- ``deleted``
+- ``absent``
+- ``present``
+
+
+``error-messages``
+------------------
+
+List of any error messages that were captured or produced while achieving a
+result.
+
+
+``exception``
+-------------
+
+An exception that occurred while achieving the reported result.
```

Comment:
```suggestion
A ``CapturedException`` that occurred while achieving the reported result. If provided, ``get_status_dict`` will fill in the ``exception_traceback`` field automatically.
```

Reply:
Is that true? It seems the code takes any exception. The feature is undocumented.

Reply:
Result rendering may take any, but `get_status_dict` doesn't. So, at least when taking advantage of that convenience feature `CapturedException` would be required.

Reply:
FWIW: I'm fine with declaring that `get_status_dict` *should* be able to deal with any. However, that's defining current behavior as a bug and I'd argue it's still worth documenting that feature here.

Reply:
I have no desire to declare anything. I am documenting what I think I know is going on. It seems that this feature was exactly how I had it in mind and described it (added by @yarikoptic in https://github.com/datalad/datalad/commit/b737726673c4e47ad3e15d78654eb981146bbe79), but was silently changed (now restricted to `CapturedException`) by @bpoldrack with the recent https://github.com/datalad/datalad/commit/6d4f1ebb901010725b9dba4dd4d2e75af474dbfe.